### PR TITLE
[qt5-base] fix x64-osx -> arm64-osx cross compile

### PR DESCRIPTION
**Pull request by ghost**
Make it possible to cross-compile qt5-base from x64 mac to arm64 by detecting cross compilation and passing the correct flags

- #### What does your PR fix?  
  

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  osx, Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by Lucius-Q-User** [comment]
Somewhat depends on #18630 , #18625 and  #18608

**Comment by JackBoosY** [comment]
Any progress?

**Comment by Lucius-Q-User** [comment]
Right now the problem is that fontconfig fails to cross compile from x86 to arm. I can rebase this patch, but if merged without fontconfig fixes, it is kinda useless, since the dependencies can't cross compile.

**Comment by JackBoosY** [comment]
cc @Neumann-A

**Comment by JackBoosY** [comment]
Does this PR no longer depend on other PRs? I see the checks are okay now.

**Comment by Lucius-Q-User** [comment]
It can be merged now.

**Comment by JackBoosY** [comment]
@Neumann-A Do you against this changes?

File: ports/qt5-base/cmake/configure_qt.cmake [hunk 1]
```diff
@@ -4,11 +4,17 @@ function(configure_qt)
...
         message(FATAL_ERROR "configure_qt requires a TARGET_PLATFORM argument.")
     endif()
-    
+
     if(DEFINED _csc_HOST_PLATFORM)
         list(APPEND _csc_OPTIONS -platform ${_csc_HOST_PLATFORM})
     endif()
-    
+
+    if(VCPKG_TARGET_IS_OSX AND NOT ("${TARGET_TRIPLET}" STREQUAL "${HOST_TRIPLET}"))
```
Comment (JackBoosY) [inline ports/qt5-base/cmake/configure_qt.cmake:12]:
```suggestion
    if(VCPKG_TARGET_IS_OSX AND VCPKG_CROSSCOMPILING)
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/qt5-base/portfile.cmake [hunk 2]
```diff
@@ -91,13 +92,6 @@ qt_download_submodule(  OUT_SOURCE_PATH SOURCE_PATH
                             patches/limits_include.patch       # Add missing includes to build with gcc 11
                     )
 
-# Remove vendored dependencies to ensure they are not picked up by the build
-foreach(DEPENDENCY zlib freetype harfbuzz-ng libjpeg libpng double-conversion sqlite pcre2)
-    if(EXISTS ${SOURCE_PATH}/src/3rdparty/${DEPENDENCY})
-        file(REMOVE_RECURSE ${SOURCE_PATH}/src/3rdparty/${DEPENDENCY})
-    endif()
-endforeach()
-#file(REMOVE_RECURSE ${SOURCE_PATH}/include/QtZlib)
```
Comment (Neumann-A) [inline ports/qt5-base/portfile.cmake:100]: Why is this removed?

**Comment by Neumann-A** [comment]
> @Neumann-A Do you against this changes?

only against the changes which removes the removal of the third party deps. These folders are removed to not accidentally build them.

**Comment by Lucius-Q-User** [comment]
Qt's build system really wants to use vendored zlib to compile the bootstrap binary (the actual arm64 binary uses system zlib). And i haven't been able to convince it to use system zlib for that.

**Comment by JackBoosY** [comment]
Any progress?
Can you please merge to master?

File: ports/qt5-base/portfile.cmake (hunk 2 above)
Comment (Lucius-Q-User) [reply ports/qt5-base/portfile.cmake]: Zlib needs to be removed since the cross-compile driver really wants to use vendored zlib. Rest dont need to be there, so i put it back.

Comment (dg0yt) [reply ports/qt5-base/portfile.cmake]: If you need a patch to change this (and cannot use host triplet tools):
https://github.com/OpenOrienteering/superbuild/blob/qt-5.15-openorienteering/qtbase/patches/0001-Use-system-zlib-for-host-tools.patch
It may need adjustment for windows triplets.

File: ports/qt5-base/patches/host-zlib.patch [hunk 3]
```diff
@@ -0,0 +1,35 @@
+--- a/mkspecs/features/qt_configure.prf
++++ b/mkspecs/features/qt_configure.prf
+@@ -2161,7 +2161,10 @@
+     for (type, $${currentConfig}.files._KEYS_) {
+         contains(type, ".*Pro") {
+             for (k, $${currentConfig}.output.$${type}.assign._KEYS_): \
++            {
+                 $${currentConfig}.output.$$type += "$$k = $$eval($${currentConfig}.output.$${type}.assign.$$k)"
++                equals(k, "QMAKE_LIBS_ZLIB"): $${currentConfig}.output.$$type += "host_build:qtConfig(system-zlib): $$k = -lz"
```
Comment (dg0yt) [inline ports/qt5-base/patches/host-zlib.patch:9]: Instead of `-lz`, we need the value of `ZLIB_LIBS`.

File: ports/qt5-base/portfile.cmake [hunk 4]
```diff
@@ -81,6 +81,8 @@ qt_download_submodule(  OUT_SOURCE_PATH SOURCE_PATH
                             patches/mysql_plugin_include.patch #Fix include path of mysql plugin
                             patches/mysql-configure.patch      #Fix mysql project
                             patches/cocoa.patch                #Fix missing include on macOS Monterrey, https://code.qt.io/cgit/qt/qtbase.git/commit/src/plugins/platforms/cocoa?id=dece6f5840463ae2ddf927d65eb1b3680e34a547
+                            patches/arm64-osx.patch            #Make it build on arm64 macos
+			    patches/host-zlib.patch            #Make the build system use host zlib instead of vendored one
```
Comment (dg0yt) [inline ports/qt5-base/portfile.cmake:85]: Oops, intentation.

**Comment by JackBoosY** [comment]
Convert this PR to draft since there is no progress.

**Comment by JackBoosY** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.
